### PR TITLE
Fix TFPolicyGraph.build_apply_op when there is no apply_op

### PR DESCRIPTION
**Pull request by vuoristo**
## What do these changes do?

When using a custom model with `DQNAgent` setting `hiddens=[]` and `dueling=False` will result in empty `_grads_and_vars` in `TFPolicyGraph`, which in turn causes an error in `build_apply_op`.

This change skips `TFPolicyGraph.build_apply_op` if `_grads_and_vars` is empty.

```
Traceback (most recent call last):
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/tune/ray_trial_executor.py", line 200, in start_trial
    self._start_trial(trial, checkpoint)
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/tune/ray_trial_executor.py", line 129, in _start_trial
    or trial._checkpoint.value is not None)
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/tune/ray_trial_executor.py", line 104, in _setup_runner
    return cls.remote(config=trial.config, logger_creator=logger_creator)
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/actor.py", line 234, in remote
    return self._remote(args=args, kwargs=kwargs)
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/actor.py", line 300, in _remote
    *copy.deepcopy(args), **copy.deepcopy(kwargs))
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/rllib/agents/agent.py", line 283, in __init__
    Trainable.__init__(self, config, logger_creator)
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/tune/trainable.py", line 88, in __init__
    self._setup(copy.deepcopy(self.config))
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/rllib/agents/agent.py", line 380, in _setup
    self._init(self.config, self.env_creator)
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/rllib/agents/dqn/dqn.py", line 207, in _init
    env_creator, self._policy_graph)
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/rllib/agents/agent.py", line 531, in make_local_evaluator
    extra_config or {}))
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/rllib/agents/agent.py", line 748, in _make_evaluator
    remote_env_batch_wait_ms=config["remote_env_batch_wait_ms"])
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/rllib/evaluation/policy_evaluator.py", line 299, in __init__
    self._build_policy_map(policy_dict, policy_config)
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/rllib/evaluation/policy_evaluator.py", line 699, in _build_policy_map
    policy_map[name] = cls(obs_space, act_space, merged_conf)
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/rllib/agents/dqn/dqn_policy_graph.py", line 461, in __init__
    update_ops=q_batchnorm_update_ops)
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/rllib/evaluation/tf_policy_graph.py", line 153, in __init__
    self._grads_and_vars)
  File "/Users/ristovuorio/projects/ray_doodle/ray/python/ray/rllib/evaluation/tf_policy_graph.py", line 297, in build_apply_op
    global_step=tf.train.get_or_create_global_step())
  File "/Users/ristovuorio/miniconda3/envs/ray_fiddle/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py", line 570, in apply_gradients
    raise ValueError("No variables provided.")
ValueError: No variables provided.
```

The code below demonstrates the error at 798944fb.

``` python
from tensorflow.keras.layers import Dense

import ray
from ray.rllib.models import ModelCatalog, Model
from ray.rllib.agents.dqn import DQNAgent
from ray.tune import run_experiments


class DemoNN(Model):
    def _build_layers_v2(self, input_dict, num_outputs, options):
        x = input_dict["obs"]
        x = Dense(1)(x)
        x = Dense(2)(x)

        return x, x


if __name__ == "__main__":
    ray.init(local_mode=True)
    ModelCatalog.register_custom_model("demo_nn", DemoNN)

    run_experiments({
        "demo": {
            "run": DQNAgent,
            "env": "CartPole-v0",
            "config": {
                "model": {
                    "custom_model": "demo_nn",
                },
                "hiddens": [],
                "dueling": False,
            },
        },
    })
```

### CODE REVIEW

File: python/ray/rllib/evaluation/tf_policy_graph.py [hunk 1]
```diff
@@ -148,9 +148,12 @@ def __init__(self,
...
                 self._update_ops))
-        with tf.control_dependencies(self._update_ops):
-            self._apply_op = self.build_apply_op(self._optimizer,
-                                                 self._grads_and_vars)
+        if self._grads_and_vars:
+            with tf.control_dependencies(self._update_ops):
+                self._apply_op = self.build_apply_op(self._optimizer,
+                                                     self._grads_and_vars)
+        else:
+            self._apply_op = None
```
Comment (ericl) [inline python/ray/rllib/evaluation/tf_policy_graph.py]: could do tf.no_op()?

Comment (vuoristo) [reply python/ray/rllib/evaluation/tf_policy_graph.py]: yeah that would make sense

**Comment by ericl** [comment]
Actually, I'm a bit confused. Why aren't there any variables? Optimization can't work if they aren't found right?

**Comment by vuoristo** [comment]
> Actually, I'm a bit confused. Why aren't there any variables? Optimization can't work if they aren't found right?

**Comment by vuoristo** [comment]
> Actually, I'm a bit confused. Why aren't there any variables? Optimization can't work if they aren't found right?

Yeah, you are right. Nothing is getting trained in the demonstration code. I dug into it a bit more and looks like the problem might be with using keras layers. I'll update here once I've figured this out.

**Comment by vuoristo** [comment]
According to [this tensorflow issue](https://github.com/tensorflow/tensorflow/issues/20842), the variable reuse behavior is different for `tf.keras.layers` than the other layer apis. So the variable scopes in `DQNPolicyGraph` don't actually cause the variables of the keras layers to be reused, and that breaks the gradient computation leading to empty `_grads_and_vars`.

If I understood correctly, going forward,`tf.keras` is the preferred way of creating layers in tensorflow, so maybe the reuse logic needs to be reworked, but that seems like a larger decision than this pr, so I'll just close it.

**Comment by ericl** [comment]
Looks related to https://github.com/ray-project/ray/issues/4134

**Pull request closed without merging**